### ISSUE TITLE
Error in `posix_memalign` on `x86_64-linux-musl`

### ISSUE THREAD

**Issue by visr** [JuliaPackaging/BinaryBuilder.jl#387]
This comes from here: https://travis-ci.org/JuliaGeo/GEOSBuilder/builds/442642120
Interestingly this only happens on `x86_64-linux-musl`.

Since it seems to me it's having trouble compiling something from stdlib, I'm raising it as an issue here. Let me know if I can provide any more information.

```
libtool: compile:  /opt/x86_64-linux-musl/bin/x86_64-linux-musl-g++ -std=c++11 -DHAVE_CONFIG_H -I. -I../../../include -I../../../include/geos -I../../../include -DGEOS_INLINE -pedantic -Wall -Wno-long-long -ffloat-store -DUSE_UNSTABLE_GEOS_CPP_API -g -O2 -MT SimplePointInAreaLocator.lo -MD -MP -MF .deps/SimplePointInAreaLocator.Tpo -c SimplePointInAreaLocator.cpp -o SimplePointInAreaLocator.o >/dev/null 2>&1
In file included from /opt/x86_64-linux-musl/lib/gcc/x86_64-linux-musl/4.8.5/include/xmmintrin.h:38:0,
                 from /opt/x86_64-linux-musl/lib/gcc/x86_64-linux-musl/4.8.5/include/x86intrin.h:34,
                 from /opt/x86_64-linux-musl/x86_64-linux-musl/include/c++/4.8.5/x86_64-linux-musl/bits/opt_random.h:33,
                 from /opt/x86_64-linux-musl/x86_64-linux-musl/include/c++/4.8.5/random:51,
                 from /opt/x86_64-linux-musl/x86_64-linux-musl/include/c++/4.8.5/bits/stl_algo.h:65,
                 from /opt/x86_64-linux-musl/x86_64-linux-musl/include/c++/4.8.5/algorithm:62,
                 from IndexedPointInAreaLocator.cpp:29:
/opt/x86_64-linux-musl/lib/gcc/x86_64-linux-musl/4.8.5/include/mm_malloc.h:34:64: error: declaration of 'int posix_memalign(void**, size_t, size_t) throw ()' has a different exception specifier
 extern "C" int posix_memalign (void **, size_t, size_t) throw ();
                                                                ^
In file included from /opt/x86_64-linux-musl/x86_64-linux-musl/include/c++/4.8.5/cstdlib:72:0,
                 from /opt/x86_64-linux-musl/x86_64-linux-musl/include/c++/4.8.5/ext/string_conversions.h:41,
                 from /opt/x86_64-linux-musl/x86_64-linux-musl/include/c++/4.8.5/bits/basic_string.h:2815,
                 from /opt/x86_64-linux-musl/x86_64-linux-musl/include/c++/4.8.5/string:52,
                 from ../../../include/geos/geom/Coordinate.h:24,
                 from ../../../include/geos/geom/Envelope.h:25,
                 from ../../../include/geos/geom/Geometry.h:37,
                 from IndexedPointInAreaLocator.cpp:17:
/opt/x86_64-linux-musl/x86_64-linux-musl/sys-root/usr/include/stdlib.h:99:5: error: from previous declaration 'int posix_memalign(void**, size_t, size_t)'
 int posix_memalign (void **, size_t, size_t);
     ^
make[5]: *** [Makefile:497: IndexedPointInAreaLocator.lo] Error 1
make[5]: *** Waiting for unfinished jobs....
```

**Comment by staticfloat** [comment]
Googling the error message, I stumble across this: https://github.com/android-ndk/ndk/issues/91

This makes me think that maybe, just maybe, there is a bug in the GCC header files and `musl` triggers it.  Have you tried compiling against `musl` with GCC 7 instead of GCC 4?  You can do that by specifying your `Platform` object as something like `Linux(:x86_64; libc=:musl, compiler_abi=CompilerABI(:gcc7))`.

**Comment by stevengj** [comment]
I just hit this with ZMQBuilder.   It's kind of annoying not to be able to use `platforms = supported_platforms()`, since specifying the platforms manually is hard to maintain … I wonder if there is another workaround?

**Comment by stevengj** [comment]
Also, wasn't the point of using gcc 4 to allow it to work using older `libstdc++`, so we lose that if we use gcc 7?

**Comment by stevengj** [comment]
According to https://patchwork.ozlabs.org/patch/544393/, this only causes an error with `-std=c++11
-pedantic-errors`, so if we make sure to compile without `-pedantic-errors` it should be fine?  It doesn't seem like the ZMQ build is using this option, but I will dig to see if there is an equivalent flag that I should disable.

**Comment by stevengj** [comment]
It seems that there is an error for these conflicts even without `-Werror` or `-pedantic`, unfortunately.   However, it looks like the `-fms-extensions` flag [changes this error to a warning](http://lists.llvm.org/pipermail/llvm-bugs/2015-October/042645.html), as a workaround for Microsoft headers that contain a similar bug.

**Comment by stevengj** [comment]
@staticfloat, would it be possible to patch `/opt/x86_64-linux-musl/lib/gcc/x86_64-linux-musl/4.8.5/include/mm_malloc.h` in the build image?

**Comment by staticfloat** [comment]
>  It's kind of annoying not to be able to use platforms = supported_platforms(), since specifying the platforms manually is hard to maintain … I wonder if there is another workaround?

Yes, there is.  You can write a function that translates the `musl` platform to compile with GCC 7.  [Here's an example](https://github.com/JuliaPackaging/Yggdrasil/blob/8215e3669f740d6b369f1ba9584e93e587d91d12/Pixman/build_tarballs.jl#L31-L36) that uses GCC 7 only for ARM, pay extra attention to the code block at the end, which is used to make the generated `build.jl` think the binary is compatible with any Julia, which is true for pixman (because it is all C code, not C++ code).  For ZMQ, you may actually be limited to Julia build with GCC 7.  (If it actually contains C++, perhaps the best thing to do is to just build for GCC 7 and GCC 8 on musl explicitly, and just not support GCC 4 on musl).

> would it be possible to patch `/opt/x86_64-linux-musl/lib/gcc/x86_64-linux-musl/4.8.5/include/mm_malloc.h` in the build image?

I am pretty hesitant to patch the headers; you yourself can of course do so as part of your build recipe, but unless the patch is something that the musl developers themselves want to apply, I'm worried that it will break something else somewhere.

**Comment by stevengj** [comment]
> you yourself can of course do so as part of your build recipe

The build recipe has write permissio to `/opt`?

**Comment by staticfloat** [comment]
Yes; some build systems do weird things like install stuff to `/usr/local`.  The build environment runs everything within an overlay filesystem that will throw away changes you make to anything outside of `$WORKSPACE/destdir` once the build script ends; but during the build, the world is your oyster; you are `root` and can do anything except some kernel calls that could modify the state of the machine outside of the build environment (for example, adding new kernel modules).
